lib/src/Uncertainty/Distribution/ConditionalDistribution.cxx

```diff
@@ -528,33 +553,42 @@ Point ConditionalDistribution::computeExpectation(const Function & f,
   }
   // Second, the discrete components
   // For each combination of the discrete components
+  Sample currentThetas(0, theta);
   for (UnsignedInteger i = 0; i < discreteAtomsNumber; ++i)
   {
     Point currentTheta(theta);
     // Get the discrete values
-    const Point discreteNode(discreteNodes_[i]);
     Bool rejectNode;
     for (UnsignedInteger j = 0; j < discreteDimension; ++j)
     {
-      const Scalar value = discreteNode[j];
+      const Scalar value = discreteNodes_(i, j);
       currentTheta[discreteMarginalsIndices_[j]] = value;
       rejectNode = (value > thetaStar[i] + epsilon);
       if (rejectNode) break;
     }
     // Skip the current integration point if the current sub-manifold is outside of the integration region
-    if (rejectNode) continue;
-    // Now, complete by the discretization of the continuous part using Gauss integration
-    for (UnsignedInteger j = 0; j < continuousAtomsNumber; ++j)
+    if (!rejectNode)
     {
-      // Complete the filling of theta using the Gauss integration node
-      const Point continuousNode(continuousNodes_[j]);
-      for (UnsignedInteger k = 0; k < continuousDimension; ++k)
-        currentTheta[continuousMarginalsIndices_[k]] = continuousLowerBounds_[k] + 0.5 * (1.0 + continuousNode[k]) * (subPoint[k] - continuousLowerBounds_[k]);
```

Comment:
@regislebrun I rewrite this portion of code to perform the same computations as before (if there are changes, it is a bug), but I find it very suspicious.  I would expect to reset currentTheta on its value before loop on j, something like
```
 const Point discreteTheta(currentTheta);
 for (UnsignedInteger j = 0; j < continuousAtomsNumber; ++j)
 {
   const Point continuousNode(continuousNodes_[j]);
   for (UnsignedInteger k = 0; k < continuousDimension; ++k)
     discreteTheta[continuousMarginalsIndices_[k]] = continuousLowerBounds_[k] + 0.5 * (1.0 + continuousNode[k]) * (subPoint[k] - continuousLowerBounds_[k]);
     const Scalar w = conditioningDistribution_.computePDF(discreteTheta) * continuousWeights_[j];
   ...
 }
```

Reply:
@dbarbier Ok I will check this tomorrow. The Bayesian part of the distributions (ConditionalDistribution, BayesDistribution and PosteriorDistribution) is at an early stage of validation, and should be refactored. The main lines of the algorithms are ok in my view, but there is still a lot to do to get something efficient and robust. Thanks for the effort.

Reply:
@dbarbier The ConditionalDistribution::computeConditionalDistribution() method is strange as a whole. To be honest, I don't even remember its precise role, the comment being much too laconic. A good thing would be to allow the user to choose what kind of WeightedExperiment she want to use for the discretization. The default GaussProductExperiment is good only when the conditioned distribution depends smoothly on its parameters, which is not always the case (eg: a Uniform with random bounds). In this case, a LowDiscrepancyExperiment would do a better job.

Reply:
@regislebrun yes but in an other PR. The objective here is to improve only performance and this improvement is required whatever the used WeightedExperiment